### ISSUE TITLE
Non-disruptive certificate rotation

### ISSUE THREAD

**Issue by guydc** [envoyproxy/gateway#4891]
*Description*:
Currently, we have MTLS connections for:
- Envoy Gateway <> Envoy
- Envoy <> Rate Limit server

Envoy Gateway generates client and server certificates for all of the above components and typically provides them as mounted secrets to relevant pods. A job that runs in the helm pre-install and pre-upgrade hooks is responsible for rotation. 

It is a common security practice to use short-lived certificates that are rotated frequently. In Envoy Gateway, CA certificates and leaf certificates are handled with the same level of security (storage, access, ... ), and should both be rotated. 

To support frequent and non-disruptive rotation, the following is required:
- All components are capable of dynamically reloading certificates when they change. 
- Certificates are rotated in a backwards-compatible manner. For example, a new trusted CA is added and the old CA is removed only after all components were able to load the latest CA/leaf certificate. This is especially relevant for data-plane components like Rate Limit where client-facing failures may occur if Envoy and Rate Limit are not both synchronized on the latests CA/leaf certificates. 

Currently, these requirements are not met:
- Dynamic reloading:
  - Envoy Gateway supports re-reading the TLS Key, Cert and CA Cert for each connection:   https://github.com/envoyproxy/gateway/blob/23856723c58df0ad28e128481f558bce41078700/internal/crypto/cert_load.go#L54
  - Envoy supports dynamic reload of TLS Key, Cert and CA Cert through the SDS path-based reference to mounted secrets: https://github.com/envoyproxy/gateway/blob/23856723c58df0ad28e128481f558bce41078700/internal/infrastructure/common/proxy_sds.go#L10
  -  Rate Limit supports dynamic reload of TLS Key/Cert but not CA: https://github.com/envoyproxy/ratelimit/blob/6a2e8262874f012d08830cc34ba8058e66a33819/src/provider/cert_provider.go#L15
- Backwards compatible changes: Envoy Gateway currently overwrites certificates in-place, meaning the change is not backwards-compatible. https://github.com/envoyproxy/gateway/blob/23856723c58df0ad28e128481f558bce41078700/internal/cmd/certgen.go#L81

[optional *Relevant Links*:]
>Any extra documentation required to understand the issue.

**Comment by zhaohuabing** [comment]
Hi @guydc thanks for explaining the issue-it's very clear!

From my understanding, her's what we need to do:
* Keep the old CA when rotating certs
* The EG, Envoy, and ratelimit server should load both the old and new CA 

> Backwards compatible changes: Envoy Gateway currently overwrites certificates in-place, meaning the change is not backwards-compatible.

Can you help me understand why this change can cause backwards-compatiblity problem?

**Comment by kfox1111** [comment]
Maybe directly, or indirectly related to this, would be spire support (https://github.com/spiffe/spire). Should this get its own ticket?

spire can push updated certificates to the workload as it rotates things. Typically the certs may be replaced each hour, and the ca's updated daily.

Would love to be able to use spire with the envoy gateway.

**Comment by guydc** [comment]
Hi @kfox1111, I think this would deserve its own issue. Are you interested in using SPIRE only for the Gateway itself (Envoy, Envoy Gateway, Rate Limit server, ... ), or also for communication with the backends?

**Comment by arkodg** [comment]
can this be closed @guydc ?

**Comment by guydc** [comment]
> can this be closed [@guydc](https://github.com/guydc) ?

I think that the RL part still doesn't work and we don't have a test confirming the Envoy <> EG part either yet.

**Comment by yurrriq** [comment]
We're still hitting this issue with v1.6.1, for what it's worth.

**Comment by arkodg** [comment]
Can you create a new issue @yurrriq , this issue is a net new feature

**Comment by yurrriq** [comment]
Sorry, I meant we're hitting the issue OP described where the CA is replaced (I'm assuming, would need to try to trigger it again look closer) and proxy pods start complaining about invalid mTLS. Do we still want a new issue?

**Comment by yurrriq** [comment]
We "solved" the issue by renewing the proxy cert (perhaps unnecessary) and killing the oldest proxy pod, which I assume still had the old CA.
